jps/model-impl/src/com/intellij/openapi/fileTypes/impl/FileTypeAssocTable.java

```diff
@@ -84,14 +93,24 @@ public T addAssociation(@NotNull FileNameMatcher matcher, @NotNull T type) {
       Map<CharSequence, T> mapToUse = exactFileNameMatcher.isIgnoreCase() ? myExactFileNameAnyCaseMappings : myExactFileNameMappings;
       return mapToUse.put(exactFileNameMatcher.getFileName(), type);
     }
-    int i = ContainerUtil.indexOf(myMatchingMappings, p -> p.first.equals(matcher));
-    if (i == -1) {
-      myMatchingMappings.add(Pair.create(matcher, type));
-      return null;
+
+    // Since we need to sort on insertion for IJPL-149806, we allow only one thread to modify the mapping list at a time
+    myMatchingMappingsLock.lock();
+    try {
+      int i = ContainerUtil.indexOf(myMatchingMappings, p -> p.first.equals(matcher));
+      if (i == -1) {
```

Comment:
please rewrite to have only one call to sort

Reply:
But there is already only one call ever due to the return? Unless you mean something like this instead?
```java
myMatchingMappingsLock.lock();
try {
  Pair<FileNameMatcher, T> old = null;
  int i = ContainerUtil.indexOf(myMatchingMappings, p -> p.first.equals(matcher));

  if (i >= 0) {
    old = myMatchingMappings.get(i);
    myMatchingMappings.set(i, Pair.create(matcher, type));
  else {
    myMatchingMappings.add(Pair.create(matcher, type));
  }

  Collections.sort(myMatchingMappings, MY_MATCHING_MAPPINGS_COMPARATOR);
  
  return Pair.getSecond(old);
}
finally {
  myMatchingMappingsLock.unlock();
}
```
It seems to be more self-explanatory to read that way, I suppose.

Reply:
yes, precisely, that variant is better, thanks